### ISSUE TITLE
"NOSCRIPT No matching script. Please use EVAL" from Orca after Redis restarts

### ISSUE THREAD

**Issue by mattnworb** [spinnaker/spinnaker#3778]
### Issue Summary:
Orca seems unable to recover when Redis (running in a Kubernetes pod) moves to a different address

### Cloud Provider(s):
Google Cloud, Spinnaker running on Google Kubernetes Engine

### Environment:
Spinnaker 1.10.4 installed via Halyard

### Feature Area: 
Orca

### Description:
We deploy Spinnaker via Halyard to a GKE cluster.  The cluster has cluster autoscaling enabled, which will add or delete nodes as needed and move some pods around.

We've noticed that at times Orca won't process any new tasks, all pipelines will show as "Not Started" in the UI, and Orca's logs will fill with errors that look like:

```
2018-12-20 06:21:23.808 ERROR 1 --- [    scheduler-5] o.s.s.s.TaskUtils$LoggingErrorHandler    : [] Unexpected error occurred in scheduled task.
redis.clients.jedis.exceptions.JedisNoScriptException: NOSCRIPT No matching script. Please use EVAL.
	at redis.clients.jedis.Protocol.processError(Protocol.java:125)
	at redis.clients.jedis.Protocol.process(Protocol.java:161)
	at redis.clients.jedis.Protocol.read(Protocol.java:215)
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:340)
	at redis.clients.jedis.Connection.getOne(Connection.java:322)
	at redis.clients.jedis.Jedis.getEvalResult(Jedis.java:2731)
	at redis.clients.jedis.Jedis.evalsha(Jedis.java:2757)
	at redis.clients.jedis.Jedis.evalsha(Jedis.java:2751)
	at com.netflix.spinnaker.q.redis.RedisQueue.readMessageWithLock(RedisQueue.kt:323)
	at com.netflix.spinnaker.q.redis.RedisQueue.poll(RedisQueue.kt:103)
	at com.netflix.spinnaker.q.QueueProcessor.pollOnce(QueueProcessor.kt:75)
	at com.netflix.spinnaker.q.QueueProcessor.access$pollOnce(QueueProcessor.kt:37)
	at com.netflix.spinnaker.q.QueueProcessor$poll$1.invoke(QueueProcessor.kt:61)
	at com.netflix.spinnaker.q.QueueProcessor$poll$1.invoke(QueueProcessor.kt:37)
	at com.netflix.spinnaker.q.QueueProcessor.ifEnabled(QueueProcessor.kt:118)
	at com.netflix.spinnaker.q.QueueProcessor.poll(QueueProcessor.kt:57)
	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

When this happens, simply restarting Orca (by deleting it's k8s pod) seems to fix things.

At least two times when this has happened recently, we've noticed that the spin-redis pod has a shorter lifetime than the spin-orca pod, suggesting that the pod running Redis was deleted at some point (perhaps because the node was being deleted / scaled down). 

Orca's config file refers to the Redis address via a k8s Service object (name "spin-redis" in namespace "spinnaker"):

```
# in orca.yml
redis:
  connection: ${services.redis.baseUrl:redis://localhost:6379}

# in spinnaker.yml
  redis:
    port: 6379
    address: spin-redis.spinnaker
    host: 0.0.0.0
    ...
    baseUrl: redis://spin-redis.spinnaker:6379
```

Should Orca be able to gracefully detect a closed connection and reconnect?

**Comment by sbwsg** [comment]
My guess is that this commit is the source of the issue: https://github.com/spinnaker/keiko/commit/96da5891aed82e5e7d9a2574b4bc85d8af86d9e9

Looks like we switched to use `evalsha` but haven't handled the case where redis has been restarted. When that happens the `readMessageWithLockScriptSha` will no longer be valid. Redis will try to find the script cached with that sha ID, won't be able to, and will return an error.

I'm working on a fix at the moment.

**Comment by sbwsg** [comment]
A fix was patched in to 1.10 and 1.11.

**Comment by mattnworb** [comment]
@sbwsg thanks! Looks like this fix is in [1.10.9](https://www.spinnaker.io/community/releases/versions/1-10-11-changelog) since that includes Orca 1.1.4 and keiko 2.9.8 - do you know if that is correct?

**Comment by sbwsg** [comment]
Yup, that's correct!
